modules/exploits/multi/http/manageengine_auth_upload.rb

```diff
@@ -0,0 +1,427 @@
+          output = $1
+          version = [ output[0].to_i, output[2].to_i, "0", "sd" ]
+        end
+        if res.body.to_s =~ /src='\/scripts\/Login\.js\?([0-9]+)'><\/script>/     # newer builds
+          version[2] = $1.to_i
+        elsif res.body.to_s =~ /'\/style\/style\.css', '([0-9]+)'\);<\/script>/   # older builds
+          version[2] = $1.to_i
+        end
+      elsif res.body.to_s =~ /ManageEngine AssetExplorer/
+        if res.body.to_s =~ /ManageEngine AssetExplorer &nbsp;([0-9]{1}\.{1}[0-9]{1}\.?[0-9]*)/   or
+        res.body.to_s =~ /<div class="login-versioninfo">version&nbsp;([0-9]{1}\.{1}[0-9]{1}\.?[0-9]*)<\/div>/
+          output = $1
+          version = [ output[0].to_i, output[2].to_i, 0, "ae" ]
+        end
+        if res.body.to_s =~ /src="\/scripts\/ClientLogger\.js\?([0-9]+)"><\/script>/
+          version[2] = $1.to_i
+        end
+      elsif res.body.to_s =~ /ManageEngine SupportCenter Plus/
+        # All of the vulnerable sc installations are "old style", so we don't care about the major / minor version
+        version[3] = "sc"
+        if res.body.to_s =~ /'\/style\/style\.css', '([0-9]+)'\);<\/script>/
+          # ... but get the build number if we can find it
+          version[2] = $1.to_i
+        end
+      elsif res.body.to_s =~ /\/console\/ConsoleMain\.cc/
+        # IT360 newer versions
+        version[3] = "it"
+      end
+    elsif res && res.code == 302 && res.get_cookies.to_s =~ /IAMAGENTTICKET([A-Z]{0,4})/
+      # IT360 older versions, not a very good detection string but there is no alternative?
+      version[3] = "it"
+    end
+    return version
+  end
+
+
+  def check
+    version = get_version
+    # TODO: put fixed version on the two ifs below once (if...) products are fixed (sd was fixed on build 9031; ae and sc still not fixed)
+    if (version[0] <= 9 && version[0] > 4 && version[2] < 9031 && version[3] == "sd") ||
+    (version[0] <= 6 && version[2] < 99999 && version[3] == "ae") ||
+    (version[3] == "sc" && version[2] < 99999)
+      return Exploit::CheckCode::Appears
+    end
+
+    if (version[2] > 9030 && version[3] == "sd") ||
+    (version[2] > 99999 && version[3] == "ae") ||
+    (version[2] > 99999 && version[3] == "sc")
+      return Exploit::CheckCode::Safe
+    else
+      # An IT360 check always lands here, there is no way to get the version easily
+      return Exploit::CheckCode::Unknown
+    end
+  end
+
+
+  def authenticate_it360(port, path, username, password)
+    if datastore['DOMAIN_NAME'] == nil
+      vars_post = {
+          'LOGIN_ID' => username,
+          'PASSWORD' => password,
+          'isADEnabled' => "false"
+      }
+    else
+      vars_post = {
+          'LOGIN_ID' => username,
+          'PASSWORD' => password,
+          'isADEnabled' => "true",
+          'domainName' => datastore['DOMAIN_NAME']
+      }
+    end
+
+    sd_port = datastore['RPORT']
+    datastore['RPORT'] = port
```

Comment:
datastore options are designed to be configured by the user and modules shouldn't modify them. Indeed msftidy already warns:

```
$ tools/msftidy.rb modules/exploits/multi/http/manageengine_auth_upload.rb
modules/exploits/multi/http/manageengine_auth_upload.rb:172 - [INFO] datastore is modified in code with '=': datastore['RPORT'] = port
modules/exploits/multi/http/manageengine_auth_upload.rb:185 - [INFO] datastore is modified in code with '=': datastore['RPORT'] = sd_port
```

Anyway, I'm going to test the module, if works, I think I can fix, clean up a little bit code by myself and land =)